compiler/sem/semstmts.nim

```diff
@@ -1099,155 +1074,425 @@ proc semLetOrVar(c: PContext, n: PNode, symkind: TSymKind): PNode =
+      if not isDiscardUnderscore(v):
         addInterfaceDecl(c, v)
-      elif v.owner == nil:
-        v.owner = getCurrOwner(c)
-      
-      styleCheckDef(c.config, v)
-      onDef(a[j].info, v)
+    elif v.owner == nil:
+      v.owner = c.p.owner
 
-      case a.kind
-      of nkConstDef:
-        setVarType(c, v, typ)
-        v.ast = def               # no need to copy
-        b = newNodeI(nkConstDef, a.info)
-        
-        if importantComments(c.config):
-          b.comment = a.comment
-        
-        b.add newSymNode(v)
-        b.add a[1]
-        b.add copyTree(def)
+    if c.inUnrolledContext > 0:
+      v.flags.incl(sfShadowed)
+    else:
+      let shadowed = findShadowedVar(c, v)
+      if shadowed != nil:
+        shadowed.flags.incl(sfShadowed)
+        if shadowed.kind == skResult and sfGenSym notin v.flags:
+          # xxx: helpful hints like this should call a proc with a better name
+          localReport(c.config, defPart.info, reportSem(rsemResultShadowed))
+
+    if v.isError:
+      discard
+    else:
+      # xxx: this needs to be symmetric with let and var in order to unify
+      #      semantic analysis, also the fact that it likely implies bugs... :/
+      case defPart.kind
       of nkVarTuple:
-        setVarType(c, v, typ[j])
         v.ast =
-          if def[j].kind != nkExprColonExpr: def[j]
-          else: def[j][1]
-        b[j] = newSymNode(v)
+          case def[i].kind
+          of nkExprColonExpr:
+            def[i][1]
+          else:
+            def[i]
+      of nkConstDef:
+        v.ast = def
       else:
         internalError(c.config, "should never happen")
 
-    result.add b
+    # set the symbol type and add the symbol to the production
+    producedDecl[i] =
+      if v.typ != nil and not sameTypeOrNil(v.typ, vTyp):
+
+        c.config.newError(
+          r,
+          SemReport(
+            kind: rsemDifferentTypeForReintroducedSymbol,
+            sym: v,
+            typeMismatch: @[c.config.typeMismatch(
+              actual = vTyp, formal = v.typ)]))
+      else:
+        v.typ = vTyp
+
+        newSymNode2(v)
+
+    case def.kind
+    of nkEmpty:
+      let actualType = v.typ.skipTypes({tyGenericInst, tyAlias,
+                                        tyUserTypeClassInst})
+
+      producedDecl[i] =
+        if actualType.kind in {tyObject, tyDistinct} and
+            actualType.requiresInit:
+          defaultConstructionError2(c, v.typ, r)
+        else:
+          checkNilableOrError(c, producedDecl[i])
+    else:
+      if sfThread in v.flags:
+        hasError = true
+        producedDecl[^1] = c.config.newError(n[^1],
+                                             reportSem rsemThreadvarCannotInit)
+    
+    if producedDecl[i].isError:
+      hasError = true
+
+    if not hasError and sfCompileTime in v.flags:
+      var x = newNodeI(result.kind, v.info)
+      x.add producedDecl
+      vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
```

Comment:
I think this is not correct.

```nim
const a {.compileTime.} = x
```

doesn't make sense (and should be disallowed).

Reply:
Yup, autopilot, would error out in pragmas